src/tree/split_evaluator.cc

```diff
@@ -0,0 +1,251 @@
+/*!
+ * Copyright 2018 by Contributors
+ * \file split_evaluator.cc
+ * \brief Contains implementations of different split evaluators.
+ */
+#include <dmlc/registry.h>
+#include "split_evaluator.h"
+#include "param.h"
+#include "../common/common.h"
+#include "../common/host_device_vector.h"
+
+namespace dmlc {
+DMLC_REGISTRY_ENABLE(::xgboost::tree::SplitEvaluatorReg);
+} // namespace dmlc
+
+namespace xgboost {
+namespace tree {
+
+// A "meta" split evaluator that can be used for combining multiple split evaluators in a compositional manner.
+class AdditiveSplitEvaluator : public SplitEvaluator {
+ public:
+  AdditiveSplitEvaluator(std::vector<std::unique_ptr<SplitEvaluator> >& inner) {
+    for(auto& e : inner) {
+      m_inner.push_back(std::move(e));
+    }
+  }
+
+  SplitEvaluator* GetHostClone() const override {
+    std::vector<std::unique_ptr<SplitEvaluator> > clones;
+
+    for(auto& e : m_inner) {
+      clones.push_back(std::move(std::unique_ptr<SplitEvaluator>(e->GetHostClone())));
+    }
+
+    return new AdditiveSplitEvaluator(clones);
+  }
+
+  void Init(const std::vector<std::pair<std::string, std::string> >& args) override {
+    for(auto& e : m_inner) {
+      e->Init(args);
+    }
+  }
+
+  void Reset() override {
+    for(auto& e : m_inner) {
+      e->Reset();
+    }
+  }
+
+  bst_float ComputeSplitLoss(bst_uint parentID,
+                             bst_uint featureID,
+                             const GradStats& left,
+                             const GradStats& right) const override {
+    bst_float ret = 0.0f;
+
+    for(auto& e : m_inner) {
+      ret += e->ComputeSplitLoss(parentID, featureID, left, right);
```

Comment:
Could the composition multiple return values here cause NaN?

Reply:
I think this will only happen if one evaluator returns Inf and another returns -Inf, or one of them returns NaN. I can't think of a situation where a loss function would evaluate to -Inf, so this maybe isn't too much of an issue?